mcp/streamable.go

```diff
@@ -654,120 +701,370 @@ func (s *streamableClientConn) Read(ctx context.Context) (JSONRPCMessage, error)
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	case <-s.done:
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		if s.err != nil {
+			return nil, s.err // Return explicit error if connection closed due to error
+		}
 		return nil, io.EOF
 	case data := <-s.incoming:
 		return jsonrpc2.DecodeMessage(data)
 	}
 }
 
-// Write implements the [Connection] interface.
+// Write implements the [Connection] interface by enqueuing the message
+// for an asynchronous send operation. The actual sending, including retries,
+// is handled by the startMessageWriter goroutine.
 func (s *streamableClientConn) Write(ctx context.Context, msg JSONRPCMessage) error {
-	s.mu.Lock()
-	if s.err != nil {
-		s.mu.Unlock()
-		return s.err
-	}
-
-	sessionID := s._sessionID
-	if sessionID == "" {
-		// Hold lock for the first request.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-s.done:
+		s.mu.Lock()
 		defer s.mu.Unlock()
-	} else {
-		s.mu.Unlock()
-	}
-
-	gotSessionID, err := s.postMessage(ctx, sessionID, msg)
-	if err != nil {
-		if sessionID != "" {
-			// unlocked; lock to set err
-			s.mu.Lock()
-			defer s.mu.Unlock()
-		}
 		if s.err != nil {
-			s.err = err
+			return s.err
 		}
-		return err
+		return io.EOF // Connection closed
+	case s.pendingMessages <- msg: // Enqueue the message for sending
+		return nil
 	}
+}
 
-	if sessionID == "" {
-		// locked
-		s._sessionID = gotSessionID
+// startMessageWriter continuously sends messages from the pendingMessages channel,
+// applying retry logic for transient errors.
+func (s *streamableClientConn) startMessageWriter() {
+	for {
+		select {
+		case <-s.done:
+			return // Connection is closed
+		case msg := <-s.pendingMessages:
+			// Use a new context for each send attempt to allow individual retries to be cancelled
+			// if the overall connection context is cancelled.
+			// This context is cancelled by the inner goroutine once the send attempt (including retries) is done.
+			ctx, cancel := context.WithCancel(context.Background())
+
+			go func(msgToSend JSONRPCMessage) {
+				defer cancel() // Ensure context is cancelled when this goroutine finishes
+
+				currentSessionID := s.sessionID.Load().(string)
+				var lastErr error
+				for i := 0; i <= s.maxRetries; i++ {
+					// Check if the main connection has been closed during retries
+					select {
+					case <-s.done:
+						return
+					case <-ctx.Done(): // Check if the individual send context was cancelled
+						return
+					default:
+						// Continue
+					}
+
+					gotSessionID, sendErr := s.postMessage(ctx, currentSessionID, msgToSend)
```

Comment:
I think there's a fundamental problem here (fixable, but fundamental): the spec says we can _resume_ an sse stream by issuing a subsequent GET requests, but does not say that we should retry post requests.

Imagine that the server-side is a stateful server (rare but possible): we don't want to perform duplicate actions. Instead, we should only resume hanging requests once we've received the response header as well as a nonempty event ID. Then we can issue a GET with that last-event-id to resume the streaming of responses.